File: 02x10/c-014/main.go
package main

import "fmt"

func collatzRule(num int) int {
	if num%2 == 0 {
		return num / 2
	} else {
		return num*3 + 1
	}
}

func fillChain(num int, collatzMap *map[int]int, indexes *[]int) {
	stack := []int{num}
	collatz := num
	_, has := (*collatzMap)[collatz]
	for !has {
		stack = append(stack, collatz)
		collatz = collatzRule(collatz)
		_, has = (*collatzMap)[collatz]
	}
	for len(stack) > 0 {
		ln := len(stack)
		pop := stack[ln-1]
		stack = stack[:ln-1]
		moreChain := (*collatzMap)[collatzRule(pop)] + 1

		(*collatzMap)[pop] = moreChain
		*indexes = append(*indexes, pop)
	}
}

func largestChain(limit int) int {
	indexes := []int{1}
	chainLength := map[int]int{1: 1}
	for i := 2; i <= limit; i++ {
		fillChain(i, &chainLength, &indexes)
	}
	largestIndex := 0
	largestValue := 0
	for i, val := range chainLength {
		if val > largestValue {
			largestValue = val
			largestIndex = i
		}
	}
	return largestIndex
}

func main() {
	fmt.Println(largestChain(1_000_000))
}

File: 02x10/c-014/Main.java
import java.util.HashMap;
import java.util.Stack;

public class Main {

    static long collatzRule(long num) {
        return num % 2 == 0 ? num / 2 : num * 3 + 1;
    }

    static void fillChain(long nun, HashMap<Long, Long> map) {
        var stack = new Stack<Long>();
        var collatz = nun;
        stack.push(nun);

        while (!map.containsKey(collatz)) {
            stack.push(collatz);
            collatz = collatzRule(collatz);
        }
        while (!stack.empty()) {
            var pop = stack.pop();
            var moreChain = map.get(collatzRule(pop)) + 1;
            map.put(pop, moreChain);
        }
    }

    static long largestChain(long limit) {
        var chainLength = new HashMap<Long, Long>();
        chainLength.put((long)1, (long)1);
        for (var i = 2; i <= limit; i++) {
            fillChain(i, chainLength);
        }

        long largestIndex = 0;
        long largestValue = 0;
        for (var entry : chainLength.entrySet()) {
            var index = entry.getKey();
            var value = entry.getValue();
            if (value > largestValue) {
                largestValue = value;
                largestIndex = index;
            }
        }
        return largestIndex;
    }

    public static void main(String[] args) {
        System.out.println(largestChain(1_000_000));
    }
}